src/cpp/fastdds/topic/DDSSQLFilter/DDSFilterValue.cpp

```diff
@@ -0,0 +1,356 @@
+        default:
+            assert(false);
+    }
+
+    return 0;
+}
+
+/**
+ * Performs promotion to FLOAT
+ */
+static long double to_float(
+        const DDSFilterValue& value)
+{
+    switch (value.kind)
+    {
+        case DDSFilterValue::ValueKind::ENUM:
+        case DDSFilterValue::ValueKind::SIGNED_INTEGER:
+            return static_cast<long double>(value.signed_integer_value);
+
+        case DDSFilterValue::ValueKind::UNSIGNED_INTEGER:
+            return static_cast<long double>(value.unsigned_integer_value);
+
+        case DDSFilterValue::ValueKind::FLOAT_CONST:
+        case DDSFilterValue::ValueKind::FLOAT_FIELD:
+        case DDSFilterValue::ValueKind::DOUBLE_FIELD:
+        case DDSFilterValue::ValueKind::LONG_DOUBLE_FIELD:
+            return value.float_value;
+
+        // The rest of the types shall never be promoted to FLOAT
+        default:
+            assert(false);
+    }
+
+    return 0;
+}
+
+/**
+ * Performs promotion to STRING
+ */
+static void to_string_value(
+        const DDSFilterValue& in,
+        eprosima::fastrtps::string_255& out)
+{
+    assert(DDSFilterValue::ValueKind::CHAR == in.kind);
+    out.assign(&in.char_value, 1);
+}
+
+void DDSFilterValue::copy_from(
+        const DDSFilterValue& other) noexcept
+{
+    kind = other.kind;
+    switch (kind)
+    {
+        case ValueKind::BOOLEAN:
+            boolean_value = other.boolean_value;
+            break;
+
+        case ValueKind::CHAR:
+            char_value = other.char_value;
+            break;
+
+        case ValueKind::ENUM:
+        case ValueKind::SIGNED_INTEGER:
+            signed_integer_value = other.signed_integer_value;
+            break;
+
+        case ValueKind::UNSIGNED_INTEGER:
+            unsigned_integer_value = other.unsigned_integer_value;
+            break;
+
+        case ValueKind::FLOAT_CONST:
+        case ValueKind::FLOAT_FIELD:
+        case ValueKind::DOUBLE_FIELD:
+        case ValueKind::LONG_DOUBLE_FIELD:
+            float_value = other.float_value;
+            break;
+
+        case ValueKind::STRING:
+            string_value = other.string_value;
+            break;
+
+        default:
+            assert(false);
+    }
+}
+
+int DDSFilterValue::compare(
+        const DDSFilterValue& lhs,
+        const DDSFilterValue& rhs) noexcept
+{
+    if (effective_kind(lhs.kind) == effective_kind(rhs.kind))
+    {
+        switch (lhs.kind)
+        {
+            case ValueKind::BOOLEAN:
+                return lhs.boolean_value - rhs.boolean_value;
+
+            case ValueKind::CHAR:
+                return lhs.char_value - rhs.char_value;
+
+            case ValueKind::SIGNED_INTEGER:
+            case ValueKind::ENUM:
+                return compare_values(lhs.signed_integer_value, rhs.signed_integer_value);
+
+            case ValueKind::UNSIGNED_INTEGER:
+                return compare_values(lhs.unsigned_integer_value, rhs.unsigned_integer_value);
+
+            case ValueKind::FLOAT_FIELD:
```

Comment:
```suggestion
            case ValueKind::FLOAT_FIELD:
            case ValueKind::FLOAT_CONST:
```

Should we include also this case?

Reply:
I don't think so. This is the switch for operators of the same kind, and the grammar does not allow to compare constant vs constant.

Reply:
I added an explicit case with an explanation on commit 58322db